src/zarr/core/indexing.py

```diff
@@ -403,16 +403,24 @@ def __iter__(self) -> Iterator[ChunkDimProjection]:
                 dim_chunk_sel_start = self.start - dim_offset
                 dim_out_offset = 0
 
+            # Use None to indicate this selection ends at the chunk edge
+            # This is useful for is_total_slice at boundary chunks,
             if self.stop > dim_limit:
                 # selection ends after current chunk
-                dim_chunk_sel_stop = dim_chunk_len
+                dim_chunk_sel_stop = None  # dim_chunk_len
 
             else:
                 # selection ends within current chunk
-                dim_chunk_sel_stop = self.stop - dim_offset
+                if dim_chunk_ix == (self.nchunks - 1):
+                    # all of the last chunk is included
+                    dim_chunk_sel_stop = None
+                else:
+                    dim_chunk_sel_stop = self.stop - dim_offset
 
             dim_chunk_sel = slice(dim_chunk_sel_start, dim_chunk_sel_stop, self.step)
-            dim_chunk_nitems = ceildiv((dim_chunk_sel_stop - dim_chunk_sel_start), self.step)
+            dim_chunk_nitems = ceildiv(
+                ((dim_chunk_sel_stop or dim_chunk_len) - dim_chunk_sel_start), self.step
```

Comment:
This is checking if `dim_chunk_sel_stop` is None, but can either `dim_chunk_sel_stop` or `dim_chunk_len` be 0? Because that would mess up this conditional.